### PR TITLE
Replica allocation consider no-op

### PR DESCRIPTION
**Pull request by henningandersen**
This is a first step away from sync-ids. We now check if replica and
primary are identical using sequence numbers when determining where to
allocate a replica shard.

If an index is no longer indexed into, issuing a regular flush will now
be enough to ensure a no-op recovery is done.

This has the nice side-effect of ensuring that closed indices and frozen
indices choose existing shard copies with identical data over
file-overlap comparison, increasing the chance that we end up doing a
no-op recovery (only no-op and file-based recovery is supported by
closed indices).

Relates #41400 and #33888

Supersedes #41784

### CODE REVIEW

File: qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java [hunk 1]
```diff
@@ -442,6 +444,43 @@ public void testCloseIndexDuringRollingUpgrade() throws Exception {
         }
     }
 
+
+    /**
+     * We test that a closed index makes no-op replica allocation only.
+     */
+    public void testClosedIndexReplicaAllocation() throws Exception {
```
Comment (dnhatn) [inline qa/rolling-upgrade/src/test/java/org/elasticsearch/upgrades/RecoveryIT.java]: I think this test passed with the current behaviour. Can we make a small PR for this test only?

File: server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java [hunk 2]
```diff
@@ -128,6 +138,43 @@ public AllocateUnassignedDecision decideUnassignedShardAllocation(ShardRouting u
         }
     }
 
+    /**
+     * Whenever we see a new data node, we clear the information we have on primary to ensure it is at least as recent as the start
+     * of the new node. This reduces risk of making a decision on stale information from primary.
+     */
+    private void ensureAsyncFetchStorePrimaryRecency(RoutingAllocation allocation) {
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/gateway/GatewayAllocator.java]: Can you make a separate PR for this enhancement?

File: server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java [hunk 3]
```diff
@@ -380,6 +375,34 @@ private static long computeMatchingBytes(TransportNodesListShardStoreMetaData.St
...
+    }
+
+    private static boolean noopMatch(TransportNodesListShardStoreMetaData.StoreFilesMetaData primaryStore,
+                                     TransportNodesListShardStoreMetaData.StoreFilesMetaData candidateStore) {
+        // We need the maxSeqNo conditions until we support non-noop recovery for closed indices (and preferably also have
+        // retention leases in place to ensure ops based recovery will actually be performed).
+        return primaryStore.hasSeqNoInfo()
+            && primaryStore.maxSeqNo() == candidateStore.maxSeqNo()
+            && primaryStore.provideRecoverySeqNo() <= candidateStore.requireRecoverySeqNo()
+            && candidateStore.requireRecoverySeqNo() == primaryStore.maxSeqNo() + 1;
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java]: Not sure if we need the last condition?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -1291,6 +1359,59 @@ public void prepareForIndexRecovery() {
         assert currentEngineReference.get() == null;
     }
 
+    /**
+     * Finalize index recovery. Manipulate store files, clean up old files, generate new empty translog and do other
+     * housekeeping for retention leases.
+     */
+    public void finalizeIndexRecovery(CheckedRunnable<IOException> manipulateStore, long globalCheckpoint,
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Can you also make a separate PR for this enhancement?

**Review by dnhatn — COMMENTED** [review]
This looks great. Thanks @henningandersen. Would you mind splitting this PR to multiple smaller pieces?

**Comment by henningandersen** [comment]
Thanks for reviewing @dnhatn , I have marked this WIP and will split it into multiple PRs (and then close this one).

**Comment by DaveCTurner** [comment]
I have opened #46318 to track the underlying issue, and am closing this since we won't be merging it as it is now. We will certainly use it for inspiration in the work on #46318.
